### PR TITLE
[flang][cuda] Enable data transfer for descriptors

### PR DESCRIPTION
**Pull request by clementval**
Remove the TODO when data transfer is done with descriptor variables.

### CODE REVIEW

File: flang/lib/Lower/Bridge.cpp [hunk 1]
```diff
@@ -3717,8 +3717,17 @@ class FirConverter : public Fortran::lower::AbstractConverter {
                            hlfir::Entity &lhs, hlfir::Entity &rhs) {
     bool lhsIsDevice = Fortran::evaluate::HasCUDAAttrs(assign.lhs);
     bool rhsIsDevice = Fortran::evaluate::HasCUDAAttrs(assign.rhs);
-    if (rhs.isBoxAddressOrValue() || lhs.isBoxAddressOrValue())
-      TODO(loc, "CUDA data transfler with descriptors");
+
+    auto getRefIfLoaded = [](mlir::Value val) -> mlir::Value {
+      if (mlir::isa_and_nonnull<fir::LoadOp>(val.getDefiningOp())) {
```
Comment (vzakhari) [inline flang/lib/Lower/Bridge.cpp:3722]: nit: you could just use `dyn_cast_or_null` and get rid of `loadOp` declaration below.
```suggestion
      if (auto loadOp = mlir::dyn_cast_or_null<fir::LoadOp>(val.getDefiningOp())) {
```

File: flang/include/flang/Optimizer/Dialect/CUF/CUFOps.td [hunk 2]
```diff
@@ -154,13 +154,15 @@ def cuf_DataTransferOp : cuf_Op<"data_transfer", []> {
     ```
   }];
 
-  let arguments = (ins Arg<AnyReferenceLike, "", [MemWrite]>:$src,
-                       Arg<AnyReferenceLike, "", [MemRead]>:$dst,
+  let arguments = (ins Arg<AnyRefOrBoxType, "", [MemWrite]>:$src,
```
Comment (vzakhari) [inline flang/include/flang/Optimizer/Dialect/CUF/CUFOps.td:157]: Can you please clarify in the comment above what happens when the operands are box references? I assume it is following the assignment semantics, so both the LHS data and the descriptor are updated.

Comment (clementval) [reply flang/include/flang/Optimizer/Dialect/CUF/CUFOps.td:161]: Yes it should follow the assignment semantic. I added a comment in that sense.

**Comment by clementval** [comment]
@vzakhari Does it look ok for you with the changes?

**Review by vzakhari — APPROVED** [review]
Thank you, Valentin!

**Pull request merged**